Cannot get description attribute for Tree

Vaadin Framework version 8.1.0.rc1 shows exception when **Vaadin Designer** was trying to get description attribute of ```com.vaadin.ui.Tree```, because there is no implementation of ```getDescription()``` for Tree.

```
Jul 12, 2017 10:17:09 AM com.vaadin.ui.declarative.DesignAttributeHandler writeAttribute
SEVERE: Failed to invoke getter for attribute description
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.vaadin.ui.declarative.DesignAttributeHandler.writeAttribute(DesignAttributeHandler.java:213)
	at com.vaadin.ui.AbstractComponent.writeDesign(AbstractComponent.java:1281)
	at com.vaadin.ui.declarative.DesignContext.createElement(DesignContext.java:485)
	at com.vaadin.designer2.model.vaadin.DesignerDesignContext.createElement(DesignerDesignContext.java:93)
	at com.vaadin.ui.declarative.Design.createHtml(Design.java:550)
	at com.vaadin.ui.declarative.Design.write(Design.java:791)
	at com.vaadin.designer2.model.vaadin.VaadinComponentSource.write(VaadinComponentSource.java:335)
	at com.vaadin.designer2.intellij.util.VaadinUtil.createSnippetFromSource(VaadinUtil.java:145)
	at com.vaadin.designer2.intellij.util.VaadinUtil.getComponentTemplateSnippet(VaadinUtil.java:110)
	at com.vaadin.designer2.intellij.palette.PalettePanel$ComponentTransferHandler.createTransferable(PalettePanel.java:341)
	at javax.swing.TransferHandler$DragHandler.dragGestureRecognized(TransferHandler.java:1601)
	at java.awt.dnd.DragGestureRecognizer.fireDragGestureRecognized(DragGestureRecognizer.java:361)
	at javax.swing.TransferHandler$SwingDragGestureRecognizer.gestured(TransferHandler.java:1669)
	at javax.swing.TransferHandler.exportAsDrag(TransferHandler.java:744)
	at javax.swing.plaf.basic.DragRecognitionSupport.mouseDraggedImpl(DragRecognitionSupport.java:174)
	at javax.swing.plaf.basic.DragRecognitionSupport.mouseDragged(DragRecognitionSupport.java:91)
	at javax.swing.plaf.basic.BasicListUI$Handler.mouseDragged(BasicListUI.java:2757)
	at java.awt.AWTEventMulticaster.mouseDragged(AWTEventMulticaster.java:319)
	at java.awt.AWTEventMulticaster.mouseDragged(AWTEventMulticaster.java:319)
	at java.awt.AWTEventMulticaster.mouseDragged(AWTEventMulticaster.java:319)
	at java.awt.AWTEventMulticaster.mouseDragged(AWTEventMulticaster.java:319)
	at java.awt.Component.processMouseMotionEvent(Component.java:6581)
	at javax.swing.JComponent.processMouseMotionEvent(JComponent.java:3342)
	at java.awt.Component.processEvent(Component.java:6302)
	at java.awt.Container.processEvent(Container.java:2236)
	at java.awt.Component.dispatchEventImpl(Component.java:4889)
	at java.awt.Container.dispatchEventImpl(Container.java:2294)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4888)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4542)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4466)
	at java.awt.Container.dispatchEventImpl(Container.java:2280)
	at java.awt.Window.dispatchEventImpl(Window.java:2746)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:90)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.awt.EventQueue$4.run(EventQueue.java:729)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:866)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:650)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:381)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Caused by: java.lang.UnsupportedOperationException: A composite has no DOM or widget
	at com.vaadin.ui.Composite.getDescription(Composite.java:304)
	... 59 more
```

--- a/server/src/main/java/com/vaadin/ui/Composite.java
+++ b/server/src/main/java/com/vaadin/ui/Composite.java
@@ -19,10 +19,10 @@ package com.vaadin.ui;
 import java.util.Collections;
 import java.util.Iterator;
 import java.util.Objects;
-import java.util.Optional;
 
 import com.vaadin.server.ErrorMessage;
 import com.vaadin.server.Resource;
+import com.vaadin.server.SerializableFunction;
 import com.vaadin.shared.ui.ContentMode;
 import com.vaadin.shared.ui.composite.CompositeState;
 
@@ -48,7 +48,6 @@ import com.vaadin.shared.ui.composite.CompositeState;
  */
 public class Composite extends AbstractComponent implements HasComponents {
 
-    private static final String COMPOSITE_HAS_NO_DOM_OR_WIDGET = "A composite has no DOM or widget";
     /**
      * The contained component.
      */
@@ -123,7 +122,7 @@ public class Composite extends AbstractComponent implements HasComponents {
         if (getCompositionRoot() != null) {
             return Collections.singletonList(getCompositionRoot()).iterator();
         } else {
-            return Collections.<Component> emptyList().iterator();
+            return Collections.<Component>emptyList().iterator();
         }
     }
 
@@ -148,43 +147,65 @@ public class Composite extends AbstractComponent implements HasComponents {
 
     @Override
     public String getStyleName() {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        Component root = getCompositionRoot();
+        return root == null ? "" : root.getStyleName();
     }
 
     @Override
     public void setStyleName(String style) {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        getRootOrThrow().setStyleName(style);
     }
 
     @Override
     public void setStyleName(String style, boolean add) {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        getRootAbstractComponentOrThrow().setStyleName(style, add);
     }
 
     @Override
     public void addStyleName(String style) {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        getRootOrThrow().addStyleName(style);
     }
 
     @Override
     public void removeStyleName(String style) {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        getRootOrThrow().removeStyleName(style);
     }
 
     @Override
     public String getPrimaryStyleName() {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        return getRootAbstractComponentPropertyOrNull(AbstractComponent::getPrimaryStyleName);
     }
 
     @Override
     public void setPrimaryStyleName(String style) {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        getRootOrThrow().setPrimaryStyleName(style);
     }
 
     private Component getRootOrThrow() {
-        return Optional.ofNullable(getCompositionRoot())
-                .orElseThrow(() -> new IllegalStateException(
-                        "Composition root has not been set"));
+        Component root = getCompositionRoot();
+        if(root == null) throw new IllegalStateException("Composition root has not been set");
+        return root;
+    }
+
+    private AbstractComponent getRootAbstractComponentOrThrow() {
+        Component root = getRootOrThrow();
+        if (!(root instanceof AbstractComponent)) {
+            throw new IllegalStateException("Composition root is not AbstractComponent");
+        }
+        return (AbstractComponent) root;
+    }
+
+    private <T> T getRootPropertyOrNull(SerializableFunction<Component, T> getter) {
+        Component root = getCompositionRoot();
+        return root == null ? null : getter.apply(root);
+    }
+
+    private <T> T getRootAbstractComponentPropertyOrNull(SerializableFunction<AbstractComponent, T> getter) {
+        Component root = getCompositionRoot();
+        if(root instanceof AbstractComponent) {
+            return getter.apply((AbstractComponent) root);
+        }
+        return null;
     }
 
     @Override
@@ -249,84 +270,82 @@ public class Composite extends AbstractComponent implements HasComponents {
 
     @Override
     public void setId(String id) {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        getRootOrThrow().setId(id);
     }
 
     @Override
     public String getId() {
-        // Design.read relies on being able to call this
-        return null;
+        return getRootPropertyOrNull(Component::getId);
     }
 
     @Override
     public void setDebugId(String id) {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        getRootAbstractComponentOrThrow().setDebugId(id);
     }
 
     @Override
     public String getDebugId() {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        return getRootAbstractComponentPropertyOrNull(AbstractComponent::getDebugId);
     }
 
     @Override
     public String getCaption() {
-        // Design.read relies on being able to call this
-        return null;
+        return getRootPropertyOrNull(Component::getCaption);
     }
 
     @Override
     public void setCaption(String caption) {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        getRootOrThrow().setCaption(caption);
     }
 
     @Override
     public void setCaptionAsHtml(boolean captionAsHtml) {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        getRootAbstractComponentOrThrow().setCaptionAsHtml(captionAsHtml);
     }
 
     @Override
     public boolean isCaptionAsHtml() {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        return getRootAbstractComponentPropertyOrNull(AbstractComponent::isCaptionAsHtml);
     }
 
     @Override
     public Resource getIcon() {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        return getRootPropertyOrNull(Component::getIcon);
     }
 
     @Override
     public void setIcon(Resource icon) {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        getRootOrThrow().setIcon(icon);
     }
 
     @Override
     public String getDescription() {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        return getRootOrThrow().getDescription();
     }
 
     @Override
     public void setDescription(String description) {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        getRootAbstractComponentOrThrow().setDescription(description);
     }
 
     @Override
     public void setDescription(String description, ContentMode mode) {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        getRootAbstractComponentOrThrow().setDescription(description, mode);
     }
 
     @Override
     public ErrorMessage getErrorMessage() {
-        return null;
+        return getRootAbstractComponentPropertyOrNull(AbstractComponent::getErrorMessage);
     }
 
     @Override
     public ErrorMessage getComponentError() {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        return getRootAbstractComponentPropertyOrNull(AbstractComponent::getComponentError);
     }
 
     @Override
     public void setComponentError(ErrorMessage componentError) {
-        throw new UnsupportedOperationException(COMPOSITE_HAS_NO_DOM_OR_WIDGET);
+        getRootAbstractComponentOrThrow().setComponentError(componentError);
     }
 
 }

--- a/server/src/test/java/com/vaadin/tests/components/TreeTest.java
+++ b/server/src/test/java/com/vaadin/tests/components/TreeTest.java
@@ -1,17 +1,25 @@
 package com.vaadin.tests.components;
 
-import org.junit.Assert;
-import org.junit.Test;
-
 import com.vaadin.data.TreeData;
 import com.vaadin.data.provider.TreeDataProvider;
 import com.vaadin.event.CollapseEvent;
 import com.vaadin.event.CollapseEvent.CollapseListener;
 import com.vaadin.event.ExpandEvent;
 import com.vaadin.event.ExpandEvent.ExpandListener;
+import com.vaadin.server.ThemeResource;
 import com.vaadin.ui.Tree;
+import org.junit.Assert;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
 
 public class TreeTest {
+
+    public static final String TEST_CAPTION = "test caption";
+    public static final String TEST_DESCRIPTION = "test description";
+    public static final String TEST_RESOURCE_ID = "nothing.gif";
 
     private static class TreeCollapseExpandListener
             implements ExpandListener<String>, CollapseListener<String> {
@@ -26,7 +34,7 @@ public class TreeTest {
 
         @Override
         public void itemCollapse(CollapseEvent<String> event) {
-            Assert.assertEquals("Source component was incorrect", tree,
+            assertEquals("Source component was incorrect", tree,
                     event.getComponent());
             Assert.assertFalse("Multiple collapse events", collapsed);
             collapsed = true;
@@ -34,7 +42,7 @@ public class TreeTest {
 
         @Override
         public void itemExpand(ExpandEvent<String> event) {
-            Assert.assertEquals("Source component was incorrect", tree,
+            assertEquals("Source component was incorrect", tree,
                     event.getComponent());
             Assert.assertFalse("Multiple expand events", expanded);
             expanded = true;
@@ -65,12 +73,27 @@ public class TreeTest {
 
         Assert.assertFalse(listener.isExpanded());
         tree.expand("Foo");
-        Assert.assertTrue("Item not expanded", tree.isExpanded("Foo"));
-        Assert.assertTrue("Expand event not fired", listener.isExpanded());
+        assertTrue("Item not expanded", tree.isExpanded("Foo"));
+        assertTrue("Expand event not fired", listener.isExpanded());
         Assert.assertFalse(listener.isCollapsed());
         tree.collapse("Foo");
         Assert.assertFalse("Item not collapsed", tree.isExpanded("Foo"));
-        Assert.assertTrue("Collapse event not fired", listener.isCollapsed());
+        assertTrue("Collapse event not fired", listener.isCollapsed());
     }
 
+    @Test
+    public void testComponentProperties() {
+        Tree<String> tree = new Tree<>();
+        tree.setCaption(TEST_CAPTION);
+        tree.setDescription(TEST_DESCRIPTION);
+        tree.setIcon(new ThemeResource(TEST_RESOURCE_ID));
+
+        assertEquals(TEST_CAPTION,tree.getCaption());
+        assertEquals(TEST_DESCRIPTION,tree.getDescription());
+        assertEquals(TEST_RESOURCE_ID,tree.getIcon().toString());
+
+        assertFalse(tree.isCaptionAsHtml());
+        tree.setCaptionAsHtml(true);
+        assertTrue(tree.isCaptionAsHtml());
+    }
 }

